Proposal: Document defining Gutenberg Principles

nerrad opened the pull request:
I've been involved with the Gutenberg project for a few years in different contexts: contributing directly to the codebase, participating in conversations around various interfaces, consuming what others have written about the project, writing plugins and blocks extending Gutenberg, consulting with teams working on it, and teaching others various aspects of the project. While there are various approaches to Gutenberg that are indirectly "picked up" by those working on it extensively as a by product of being "in the trenches", I have found that there's no real authoritative source for describing the underlying principles guiding Gutenberg development and design.

This lack manifests itself with some difficulty when trying to:

- communicate why certain architectural decisions are made in interfaces built.
- understand from a business perspective why Gutenberg is beneficial and why its approach should be considered in building for the WordPress ecosystem.
- help third parties who will be building extensions on top of the various GB interfaces we build and why things work “differently” than what they are used to in the WordPress ecosystem.

While there are a number of posts spread through various blogs and sources that articulate pieces of “Gutenberg principles”, to date there really isn’t a cohesive document than serves as an “official” repository of this guiding knowledge.

Having official principles guiding Gutenberg development can help in the following ways:

- Reviewers have something to help use when assessing the work in submitted PRs in the Gutenberg project.
- Guiding architectural discussions and decisions around changes or improvements in general design (both user and code). Besides having a clearer idea around the “yes” things we do, it also makes it much easier to identify why we might say “no” to an idea or approach.
- Communicating with the wider development and WordPress business community that might have competing interests for how they feel Gutenberg “should be done”. This provides something to align their own work with.

I believe there is an underlying set of ideas and principles that are guiding Gutenberg development and design whether it's communicated officially or not.

Given the above, and my own understanding of the Gutenberg project from my participation, observation, and usage, I'm submitting a proposed document with what I think some of the principles are guiding GB development and design.  Some additional considerations:

- I'd love to have more variety and diversity represented in resources that complement the principles outlined in the doc before publishing. So please pop in some links for consideration if anyone knows of any.
- Related to the above, I think it'd be beneficial to consider this as an evolving doc. Even after publishing, we may find that certain principles can be refined or additional ones pop up. I also think it's beneficial to have this as an _entry point_ type of document that inspires folks to dive deeper into the implications of these principles in code, design or other posts published in our ecosystem. Ideally we could link to these various voices in the WordPress ecosystem from this doc as well over time.
- More than anything, this is intended to kickstart discussion around having these principles articulated and published. Whether or not the final form of this document ends up being what I've introduced here is less important to me than having _something_ in place.

nerrad committed:
add doc on Gutenberg Principles

nerrad committed:
fix typo with markdown emphasis

draganescu commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+Each of the words in this principle are deliberate:
+
+
+### Delightful
```
As much as I love this word, it's just not a good principle because it is so subjective. It has also been confiscated by commercial projects.

nerrad replied:
Is it primarily the specific word you object to or also the meaning behind the choice of the word (and the intent of the principle)? If just the word itself, do you have any suggestions about what could be used instead?

tellthemachines replied:
Delightful does sound a bit like marketing-speak. It's not so much the subjectivity as the underlying idea that we're somehow directing the user's feelings about the product that makes me uneasy.

The word "transparent" popped into my head when reading `The user interfaces should “get out of the way” and support the user`. Transparency both in the sense of clarity (it's clear how to use the interface, thus "supporting" the user instead of making things difficult) but also in the sense of invisibility (it gets "out of the way" so users don't even notice it while working).

nerrad replied:
You both have shared perspectives on the word "delightful" that I didn't even think of. I was thinking of it in the context of aiming for user experiences that elicit "surprised joy" (my interpretation of "delight"). I also don't think we can escape the reality that we indeed _do_ influence users feelings about the product when building it.  This principle is aimed at thinking how what we build contributes to positive feelings.

With that said, I do see how **transparent** could be a good substitute here for the word delightful and moves more towards the objectiveness of a _principle_ as opposed to the subjectivity of a value or goal.

aristath replied:
+1 for `transparent` :+1:

mtias replied:
"Delightful" doesn't capture what I think this section hints at. "Transparent" is also lacking a bit in concrete meaning.

The way I'd describe this is "A canvas for expression" (more on that here: https://github.com/WordPress/gutenberg/issues/18667)

nerrad replied:
I'll see if I can capture some of what you wrote in that issue Matías - ❤️ it.

nerrad replied:
See the changes made in d6db504.  I incorporated much of what Matías articulated in #18667 and this forms the basis for the revised content of the principle which is now simply "A canvas for expression" with the supporting points of this canvas:

- Guiding users around interacting with the content
- Uniformity within diversity
- and Accessbility

The wording around accessibility has also been tweaked in f8a44af which includes not only the work around this for _creators_ of content, but also for _consumers_ of content (to address what was [brought up here](https://github.com/WordPress/gutenberg/pull/30963#discussion_r620308866)).

This also handles the concerns around the word "delightful".

youknowriad commented:
This is a great document. I don't have much to add to be honest :)

gziolo commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+
+## Progressive Complexity
+
+_“On the Layers of the Onion”_ (Matias - Gutenberg or the Ship of Theseus)
```
It isn't linked.

nerrad replied:
Addressed in 5e8d288.

gziolo commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+
+This also encompasses the ideas of _portability and freedom from lock-in_:
+
+“One’s content shouldn’t be tied to anything. This means that content shouldn’t be tied to any runtime, be it Gutenberg or WordPress in general, and that a good relationship with other editors (mobile apps, MarsEdit, etc.) should be developed.” (Portability principle in “The Language of Gutenberg - Miguel)
```
Link is missing.

nerrad replied:
Addressed in 5e8d288.

gziolo commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+
+“One’s content shouldn’t be tied to anything. This means that content shouldn’t be tied to any runtime, be it Gutenberg or WordPress in general, and that a good relationship with other editors (mobile apps, MarsEdit, etc.) should be developed.” (Portability principle in “The Language of Gutenberg - Miguel)
+
+“Adopting Gutenberg for early testing or curiosity shouldn’t be an irreversible action for one’s content. Disabling Gutenberg shouldn’t result in unreasonably altered content. The same should later apply to third-party blocks. There is a strong desire to fight lock-in, which is not new but is pervasive in a world of proprietary software and software-as-a-service.” (No commitment principle in “The ...
```
Link is missing.

nerrad replied:
Addressed in 5e8d288.

gziolo commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+
+Imagine a world where content created by Gutenberg can be consumed and read _anywhere_. This is largely why at the heart of the block structure, the format is largely an enhancement on top of HTML.
+
+“Ultimately, choosing HTML means that — as with a painting or a sculpture — the editor’s final artefact is the canonical format of the content, not a byproduct thereof.” (The Language of Gutenberg - Miguel).
```
Link is missing.

nerrad replied:
Addressed in 5e8d288.

gziolo commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+
+This manifests in letting machines do what machines are good for, and preserve content in a format that is readable to users.
+
+Finally, the principle of ubiquitous and safe adoption is found in the way Gutenberg is being developed _incrementally_. WordPress powers ~ 40% of the Web and the changes that GB brings cannot happen overnight (Gutenberg or the ship of Theseus - Matías)
```
Link is missing.

And maybe we use "over 40%" so we don't have to change it too often :trollface: 

```suggestion
Finally, the principle of ubiquitous and safe adoption is found in the way Gutenberg is being developed _incrementally_. WordPress powers over 40% of the Web and the changes that GB brings cannot happen overnight (Gutenberg or the ship of Theseus - Matías)
```

gziolo reviewed (approved):
This is a great document. @nerrad, you did an extraordinary job to put it all together in such a thoughtful way.

I left some minor thing like missing links that I could spot when browsing:

https://github.com/WordPress/gutenberg/blob/add/gb-principles-doc/docs/explanations/principles.md

It would be great to have a few more ✅ to ensure we all share the same principles 😄

nerrad committed:
Improve reference to WordPress usage stats.

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

mkaz commented on docs/README.md:
```diff
@@ -20,6 +20,9 @@ The Editor offers rich new value to users with visual, drag-and-drop creation to
 
 ## Quick links
 
+### Gutenberg Principles
+[The Gutenberg principles document](/docs/explanations/principles.md) highlights a few principles that are guiding the overall project and shape various development and design decisions.
+
```
Can we move this section into the `Contribute to the block editor` below, I don't think it should be the first Quick Link.

This is the first page that shows when viewing the [Block Editor Handbook](https://developer.wordpress.org/block-editor/) and people are going to be more interested in how to build for and extend first.

With that also said, I think a better more prominent spot would be the [repository top-level readme](https://github.com/WordPress/gutenberg/blob/trunk/README.md), as that is only showed within the repo and more likely target to developers of the project.

nerrad replied:
I'm going to go ahead and move this into the repo top-level README.md - I agree it makes more sense to target contributors. 

Given your [comment here however](https://github.com/WordPress/gutenberg/pull/30963#pullrequestreview-643571178), do you think I still need to expose this on on the built docs and if so, where do you think it should be in the docs structure?

nerrad replied:
See 219f0af (and 1a648ba where wording was changed a bit).

mkaz reviewed (commented):
Since this is a new document you need to add it to `docs/toc.json` and then `npm run docs:build` and then commit the updated `manifest.json` file. :+1: 

See the existing listings in `toc.json` for syntax.

mtias commented:
I'd like to give this a bit more thought before merging.

mtias commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+
+Extensibility interfaces in Gutenberg favor:
+
+**Modularity** - the concept of being able to combine blocks or components into different types of layout or content (see “[Embrace the modularity](https://riad.blog/2020/01/28/embrace-the-modularity/)”).
```
There's some overlap here with https://developer.wordpress.org/block-editor/explanations/architecture/key-concepts/ — it should be linked at least

nerrad replied:
See 5e744d9.

mtias commented on docs/README.md:
```diff
@@ -20,6 +20,9 @@ The Editor offers rich new value to users with visual, drag-and-drop creation to
 
 ## Quick links
 
+### Gutenberg Principles
```
I think "Guiding Principles" would be better.

mtias commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+
+
+# Gutenberg Principles
```
I think "Guiding Principles" would be better.

nerrad replied:
See 1a648ba.

mtias commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+When applied, sometimes these principles build on each other much like the idea of constructing a house with a foundation, walls and roof. However, other times the application of the principles results in something that is more similar to a tapestry where threads of each principle are embedded together and evident in the finished feature or design of the project.
+
+
+## Aim for delightful, consistent, accessible user interfaces for empowering content creation and expression.
```
I adhere to "empowering content creation and expression" but I'm not sure the other concepts capture the right concepts. 🤔

nerrad replied:
What about something like:

> The user experience is a canvas that empowers content creation and expression through intuitive and exploration friendly interfaces

nerrad replied:
See d6db504.

mtias commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+The user interfaces should “get out of the way” and support the user in creating their content. They should lead to experiences where the user (in the vernacular of Marie Kondo) _finds delight_ with each new thing they discover they can do or accomplish.
+
+
+### Consistent
```
This is not a principle I'd be comfortable stating as a primary guiding force because it's a bit more nuanced. The way I tend to describe this is more along the lines of "Providing uniformity within diversity, as learning the ingredients of the interface happens once but scales to hundreds of blocks."

This is an important design tension.

nerrad replied:
I like the concept of "providing uniformity within diversity" but I think there's still the importance here of capturing the idea of not just the _block interfaces_ but also the interfaces within the editor? Would you say this statement is a principle that also applies to the editor interfaces?

Also, is this something that you believe should be explicitly called out as it's _own_ principle, or should it be included somehow as a part of [the statement commented on here](https://github.com/WordPress/gutenberg/pull/30963#discussion_r620302895)?

nerrad replied:
See d6db504.

mtias commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+This also applies to learning _where to find_ different elements that the user might interact with in creating their content.
+
+
+### Accessible
```
This one feels a bit odd as it's more of a general principle of WordPress at wide and not specific to Gutenberg. If anything, it should build upon the two other axis of "canvas for expression" and "uniform yet diverse" to embrace the fact users interacting with the software have distinct characteristics and preferences for how to use said software.

mtias replied:
There's also something important missing here which is how the software can help users build better websites that are semantic and more accessible. That is, _respect for the viewers as well as the creators_.

nerrad replied:
> That is, respect for the viewers as well as the creators.

This is definitely something that isn't capture well in the current iteration that should be included ❤️ 

> This one feels a bit odd as it's more of a general principle of WordPress at wide and not specific to Gutenberg. If anything, it should build upon the two other axis of "canvas for expression" and "uniform yet diverse" to embrace the fact users interacting with the software have distinct characteristics and preferences for how to use said software.

While I agree it's a general principle of WordPress, I also think it's important to call out the distinctive nature of accessibility in the context of Gutenberg. I like how you reflect on describing this as software accommodating users with "distinct characteristics and preferences for how to use said software". An example of this is reflected in the numerous entry points people have to insert a block. Is there some specific concern here around the use of the word "accessible" itself here?

nerrad replied:
See f8a44af.

mtias commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+**Experimentation** - Often new APIs and interfaces clearly start as experimental while reliability and usability of different approaches are being tested. These experiments always begin with determining what problem is solved from the _creator_ perspective and seek to validate how it impacts the user workflow. The intent is to explore what generalizations of the creator's use-case are possible a ...
+
+
+## Progressive Complexity
```
This needs to be balanced a bit more with a general seek for "simplicity"

nerrad replied:
Content wise in the description of the principle, or reflected in the principle statement itself?

nerrad replied:
This one was a bit difficult for me to address. I think in general I understand your feedback to mean we want to avoid the idea of complexity as a _starting point_ for the interactions. I've made a few changes in e7b3960 which I think helps, but it still feels like it could be better expressed.

mtias commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,106 @@
+There’s also the related application here to design systems in larger organizations where the design team wants to put some restrictions on what can be modified and manipulated by the content creation team in keeping with the chosen designs behind the presentation of that content. In this case, the progressive complexity principle in Gutenberg provides the tools for “locking down” various element ...
+
+
+## Ubiquitous and Safe Adoption
```
I this is more about a profound respect for the user in an expansive sense — both in how it respects where they come from (compatibility) and aims to ensure protections for the future (the way user data is treated, for example)

nerrad replied:
So is this more a comment on the principle wording itself (that you would like changed) or on the description of the principle?

nerrad replied:
I left this as is, it's not clear to me how to improve it.

nerrad committed:
Simplify first principle and improve the description.

This moves away from using the word delightful and incorporates some language used by Matias to describe interface principles (@see https://github.com/WordPress/gutenberg/issues/18667)

nerrad committed:
Edits to references and ensure all are linked.

nerrad committed:
move link to principles doc into the top-level project README.md

nerrad committed:
Add link to key architectural concepts (impacting extensibility)

nerrad committed:
changed title of doc and tweak wording of relevant link in top-level readme

nerrad committed:
Improve wording around accessibility

nerrad committed:
demphasize “complex” in progressive complexity principle.

nerrad commented:
Sorry for the delay in addressing all the feedback. I've gone through and made changes and commented in the related threads to help highlight what was changed as a result. I think this is ready for another round of examination as folks are able.

gziolo commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,97 @@
+
+**Experimentation** - Often new APIs and interfaces clearly start as experimental while reliability and usability of different approaches are being tested. These experiments always begin with determining what problem is solved from the _creator_ perspective and seek to validate how it impacts the user workflow. The intent is to explore what generalizations of the creator's use-case are possible a ...
+
+For more information around currently built patterns for extensibility and key architectural concepts, see [this document](https://developer.wordpress.org/block-editor/explanations/architecture/key-concepts/)).
```
```suggestion
For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md)).
```

nerrad replied:
oops, thanks for making these changes to add the relative links @gziolo :)

gziolo committed:
Update docs/explanations/principles.md

gziolo commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,97 @@
+
+**Experimentation** - Often new APIs and interfaces clearly start as experimental while reliability and usability of different approaches are being tested. These experiments always begin with determining what problem is solved from the _creator_ perspective and seek to validate how it impacts the user workflow. The intent is to explore what generalizations of the creator's use-case are possible a ...
+
+For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md)).
```
```suggestion
For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md).
```

gziolo committed:
Update docs/explanations/principles.md

getsource commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,97 @@
+
+## Curated Extensibility
+
+The use of the word _curated_ expressly refers to implementing extensibility in a way that prioritizes those _using Gutenberg interfaces to create their content_ over developers integrating with the system. As a result, every extensibility interface request is filtered through the question of how it impacts creator workflows.
```
I'm having a little bit of trouble parsing this paragraph.

In "_using Gutenberg interfaces to create their content_", does `interfaces` refer to UI/user interfaces, or interfaces as a whole, including APIs?

Assuming the prior, maybe "prioritizes those using Gutenberg user interfaces to create their content"?

nerrad replied:
Good point. Fixed in 9b4eade (which also clarifies the usage of `interfaces` elsewhere in the doc).

getsource replied:
Looks good; thanks much!

nerrad committed:
Clarify usage of the word interfaces in the doc.

nerrad commented:
Just checking in to see if there's any appetite for seeing this land from other @WordPress/gutenberg-core contributors. I let this sit for a bit given the efforts around getting WordPress 5.8 released. In particular, @mtias, I addressed some feedback you had and I'm not sure you've seen it, or if you have additional comments.

I'm happy to discuss/iterate more if needed (or even close the PR if there's no consensus).

mtias commented:
Thanks for the nudge — I'm still not quite happy with some of the items and headings.

nerrad commented:
> I'm still not quite happy with some of the items and headings.

Based on this, it's not clear to me what are the next steps. I've assumed don't merge. I also do not know which items and headings are still problematic or what needs done to make them better.

mtias commented:
Sorry, I mean that some of the items in the outline feel at different levels but is not quite clear why. I don't have a concrete suggestion just yet. Perhaps it's trying to cover too much at once.

gziolo commented:
That's very unfortunate to see this PR closed with no additions to the Block Editor Handbook 😞 

I was hoping we will finally have a good reference to use when discussing new extensibility points.

getdave commented:
Agreed. How about after 5.9 we look to break this PR down into smaller documentation updates to various sections of the Block Editor Handbook? That way it will be easier to review and merge incrementally.

nerrad commented:
Sorry folks, I closed this mostly because I felt it was stalled and I really had no clear idea on what some next steps could be taken to get it to a mergable state. While I do still think it'd be really helpful to have some master doc outlining principles guiding the project, I got a sense (maybe incorrectly) that there's still not sufficient consensus behind publishing this :). That's not necessarily a bad thing, it just means there's more work to do before something like this PR can be published. 

> How about after 5.9 we look to break this PR down into smaller documentation updates to various sections of the Block Editor Handbook? That way it will be easier to review and merge incrementally.

One thing I've noticed with the feedback that has come in, is that there _are_ some principles already sprinkled through various issues and docs already. Maybe all that is needed is to capture those and pull them into an index of sorts. The main outcome I was hoping from this PR was to have a single reference to point folks to for all the reasons I outlined in the PR description.

nerrad commented on docs/explanations/principles.md:
```diff
@@ -0,0 +1,97 @@
+For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md).
+
+
+## Progressive Complexity unfolding from simple origins.
```
Still not happy with this heading. Feels too wordy and abstract.

mtias commented:
I personally don't mind this sitting open, we can come back to it, others can read it and add thoughts. Some time distance can also clarify what works and what doesn't.

senadir commented:
Now that 5.9.0 is merged, would work on this get resumed?

priethor commented:
> Should we close this PR, or is there a plan to continue working on core principles?

It's been stale for more than 3 years, I think it's fine to close it and reopen it if we get to revisit the idea.
